### PR TITLE
Replace fastaparser dependency

### PR DESCRIPTION
**Pull request by WardBrian**
This removes the GPL library `fastaparser` and replaces the functionality with a very simple version we can maintain ourselves. The basic structure of the parser is very similar to the one in the BSD-3 BioPython package ([source](https://github.com/biopython/biopython/blob/76796c970a68fa09da94dc95827b217a872004e2/Bio/SeqIO/FastaIO.py#L24)) but I don't think we need to depend on such a large package for this.

### CODE REVIEW

**Comment by gowerrobert** [comment]
I checked out this branch, ran the new test/test_fasta.py and it passed, I read the code and it  makes since, and tested the new fasta parser with other isoform files and it worked.

File: kmerexpr/fasta.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+class Sequence:
+    header: str
+    sequence: str
+
+
+def read_fasta(handle: TextIOBase) -> Generator[Sequence, None, None]:
+    """
+    Basic lazy FASTA format reader.
+
+    This assumes the only comments are headers. Sequences can be split
```
Comment (bob-carpenter) [inline kmerexpr/fasta.py:23]: I'm not sure what "only comments are headers" means.  Looking at the code, lines before the first line starting with `>` are just skipped and that after that, everything is either part of a header or part of a sequence.  Is that what this means?  If that's right, can we say that more directly?

File: requirements.txt [hunk 2]
```diff
@@ -13,4 +11,4 @@ scipy==1.7.3
 # NUMBA
```
Comment (bob-carpenter) [inline requirements.txt:11]: I'm confused.  NUMBA is BSD-2, so shouldn't the comment be about the license to match everything else?

Also, why are we using cryptography here?

For plotting, I'd like to switch to plotnine, but that's going to involve bringing @gowerrobert around, so not yet.

File: kmerexpr/transcriptome_reader.py [hunk 3]
```diff
@@ -34,17 +34,23 @@ def shred(seq, K):
...
+    float_t=np.float32,
+    int_t=np.int32,
 ):
     print("K =", K)
     print("fasta file =", fasta_file)
     print("target x file =", x_file)
     print("float type =", float_t)
     print("int type =", int_t)
-    M = 4**K
+    M = 4 ** K
```
Comment (bob-carpenter) [inline kmerexpr/transcriptome_reader.py:50]: [Comment]
I generally prefer space around operators, but this looks strange to see it around the power operator.

**Review by bob-carpenter — CHANGES_REQUESTED** [review]
Looks good.  The only real request here is in the format of fast files that will be parsed and comments in the dependencies.

I wish I could figure out how to remove the repeated code for header/seq extraction in the ad hoc parser.  It's like a repeat-until or do-while thing and it bugs me.  But I don't see an easy way to do that.  And it looks like this will deal with empty lines just fine (by stripping and appending).

**Comment by bob-carpenter** [comment]
Thanks, @gowerrobert --- I checked the code pretty closely.  I still find it annoying I can't make it simpler, but that's the nature of these ad hoc parsers and not having a general push-back mechanism.  I could shorten this one a bit, but it adds two conditionals that are redundant and requires an extra nested test per sequence, which is presumably going to make it slower.  

```python
head = ""
bases = ""
for line in handle:
    if line.startswith(">"):
        if head:
            yield Sequence(head, bases)
        head = line[1:].rstrip()
        bases = ""
    else:
        bases += line.strip()
if head:
    yield Sequence(head, bases)
```

Not worth changing, but it still bugs me I can't make this tighter.  

Now that I think about this more, I think I wrote the parser originally in C++, but then couldn't get Eigen to work on the sparse matrices (it segfaulted when scaling a short signed int number of rows).  And then just grabbed a library when we switched over.

File: kmerexpr/fasta.py (hunk 1 above)
Comment (WardBrian) [reply kmerexpr/fasta.py:23]: [Wikipedia](https://en.wikipedia.org/wiki/FASTA_format) has some examples of sequences with multiple comments, something like 

```
> Some header
> Some second comment
GATTACA
```
The idea is that the first line is the header, the second line is ignored entirely, and the third line is the sequence. 

This is valid under "the original Pearson FASTA format", but not "the NCBI FASTA specfication", according to the article. 

As far as I can tell, none of the open source parsers will accept it.

File: requirements.txt (hunk 2 above)
Comment (WardBrian) [reply requirements.txt:11]: Those were added in 253329a32dcf279202e36f591676900ffdfbc2ed a long time ago. It looks like @gowerrobert may have just dumped his pip environment into the file. 

I didn't change anything besides delete fastaparser here, so I'm not too sure about the other contents

File: kmerexpr/transcriptome_reader.py (hunk 3 above)
Comment (WardBrian) [reply kmerexpr/transcriptome_reader.py:50]: I think that was a change `black` made, I didn't do it manually

**Comment by WardBrian** [comment]
I agree that the code's structure is unsatisfying. I think most other forms have their own warts or notably different runtime behavior, like you mention with the extra test conditionals.

**Review by bob-carpenter — APPROVED** [review]

**Pull request merged**